fakestorage/upload.go

```diff
@@ -150,17 +150,54 @@ func (s *Server) insertFormObject(r *http.Request) xmlResponse {
 func (s *Server) checkUploadPreconditions(r *http.Request, bucketName string, objectName string) *jsonResponse {
 	ifGenerationMatch := r.URL.Query().Get("ifGenerationMatch")
 
-	if ifGenerationMatch == "0" {
-		if _, err := s.backend.GetObject(bucketName, objectName); err == nil {
+	if ifGenerationMatch != "" {
+		gen, err := strconv.ParseInt(ifGenerationMatch, 10, 64)
+		if err != nil {
+			return &jsonResponse{
+				status:       http.StatusBadRequest,
+				errorMessage: err.Error(),
+			}
+		}
+		_, err = s.backend.GetObjectWithGeneration(bucketName, objectName, gen)
+		if gen == 0 {
+			if err == nil {
+				return &jsonResponse{
+					status:       http.StatusPreconditionFailed,
+					errorMessage: "Precondition failed",
+				}
+			}
+		} else if err != nil {
 			return &jsonResponse{
 				status:       http.StatusPreconditionFailed,
 				errorMessage: "Precondition failed",
 			}
 		}
-	} else if ifGenerationMatch != "" || r.URL.Query().Get("ifGenerationNotMatch") != "" {
-		return &jsonResponse{
-			status:       http.StatusNotImplemented,
-			errorMessage: "Precondition support not implemented",
+		return nil
```

Comment:
random/non-blockign q: does this match the behavior of GCS when both parameters are provided. I know that providing both doesn't make sense, just wanted to confirm GCS prioritizes `ifGenerationMatch` in the presence of both (or maybe it fails with 400 Bad Request? idk)